lib/puma/systemd.rb

```diff
@@ -0,0 +1,46 @@
+# frozen_string_literal: true
+
+require 'sd_notify'
+
+module Puma
+  class Systemd
+    def initialize(events)
+      @events = events
+    end
+
+    def hook_events
+      @events.on_booted { SdNotify.ready }
+      @events.on_stopped { SdNotify.stopping }
+      @events.on_restart { SdNotify.reloading }
+    end
+
+    def start_watchdog
+      return unless SdNotify.watchdog?
+
+      ping_f = watchdog_sleep_time
+
+      log "Pinging systemd watchdog every #{ping_f.round(1)} sec"
+      Thread.new do
```

Comment:
Hmm... doesn't it defeat the purpose if you start the watchdog as a separate thread? How would that ever fail? It should either be integrated into the main event loop or at least somehow check whether the service is actually still working.

Reply:
I think that's a fair point but I wouldn't know how to do it better. Perhaps this is a good enough implementation for now and it allows for a better implementation in the future. Maybe the thread should perform a request to see if it's served? The question then becomes: which one? Should it be configurable?